src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,48 +3,189 @@
+        } else {
+            linkLast(value);
+        }
+        return true;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index < 0 || index > size) {
+            throw new ArrayIndexOutOfBoundsException("Out of bounds search index");
+        }
+        if (index == size) {
+            linkLast(value);
+        } else {
+            linkBefore(value, node(index));
+        }
     }
 
     @Override
     public boolean addAll(List<T> list) {
-        return false;
+        T[] addArray = (T[]) list.toArray();
+        int lengthOfArray = addArray.length;
+        if (lengthOfArray == 0) {
+            return false;
+        }
+        Node<T> prev;
+        prev = last;
+        for (T elementAddArray : addArray) {
+            Node<T> inputNode = new Node<>(elementAddArray, null, prev);
+            if (prev == null) {
+                first = inputNode;
+            } else {
+                prev.next = inputNode;
+            }
+            prev = inputNode;
+        }
+        size += lengthOfArray;
+        return true;
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        return node(index).element;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+        Node<T> oldNode = node(index);
+        T oldValue = oldNode.element;
+        oldNode.element = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        return unlink(node(index));
     }
 
     @Override
     public boolean remove(T t) {
+        if (t == null) {
+            for (Node<T> i = first; i != null; i = i.next) {
+                if (i.element == null) {
+                    unlink(i);
+                    return true;
+                }
+            }
+        } else {
+            for (Node<T> i = first; i != null; i = i.next) {
+                if (i.element.equals(t)) {
+                    unlink(i);
+                    return true;
+                }
+            }
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void linkFirst(T e) {
+        final MyLinkedList.Node<T> f = first;
+        final MyLinkedList.Node<T> newNode = new MyLinkedList.Node<>(e, f, null);
+        first = newNode;
+        if (f == null) {
+            last = newNode;
+        } else {
+            f.prev = newNode;
+        }
+        size++;
     }
+
+    void linkLast(T e) {
+        final Node<T> l = last;
+        final Node<T> newNode = new MyLinkedList.Node<>(e, null, l);
+        last = newNode;
+        if (l == null) {
+            first = newNode;
+        } else {
+            l.next = newNode;
+        }
+        size++;
```

Comment:
1. create informative variable names
2. the use of `final` keyword must have a reason other than just bypassing the checkstyle
3. you call this method if size != 0, can `last` be null?

Reply:
1. Done;
2. Done;
3. Similarly linkFirst.

Reply:
3. I am calling the method "linkLast" in method "add(T value, int index)" and size inside this method can be == 0 and last can be null